lib/api/include/irods/delay_rule_claim.h

```diff
@@ -0,0 +1,101 @@
+#define IRODS_DELAY_RULE_CLAIM_H
+
+/// \file
+
+#include "irods/objInfo.h"
+#include "irods/rodsDef.h" // For TIME_LEN
+
+struct RcComm;
+
+/// The input data type used to invoke #rc_delay_rule_claim.
+///
+/// \since 5.0.0
+typedef struct DelayRuleClaimInput
+{
+    /// The catalog ID of the delay rule to claim.
+    ///
+    /// Must be a non-empty string.
+    ///
+    /// \since 5.0.0
+    char rule_id[32]; // NOLINT(cppcoreguidelines-avoid-c-arrays, modernize-avoid-c-arrays)
+
+    /// The FQDN, hostname, or IP of the delay server which is attempting to claim the
+    /// delay rule.
+    ///
+    /// Must be a non-empty string.
+    ///
+    /// \since 5.0.0
+    char claim_host[300]; // NOLINT(cppcoreguidelines-avoid-c-arrays, modernize-avoid-c-arrays)
+
+    /// The PID of the delay server which is attempting to claim the delay rule.
+    ///
+    /// \since 5.0.0
+    int claim_host_pid;
+
+    /// The set of key-value pair strings used to influence the behavior of the API
+    /// operation.
+    ///
+    /// \since 5.0.0
+    struct KeyValPair cond_input;
+} delayRuleClaimInp_t;
+
+// NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+#define DelayRuleClaimInput_PI "str rule_id[32]; str claim_host[300]; int claim_host_pid; struct KeyValPair_PI;"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// Attaches information about the delay server to a delay rule in the catalog.
+///
+/// This API is designed to allow the delay server to block other delay servers from executing
+/// a delay rule.
+///
+/// Requires \p rodsadmin level privileges.
+///
+/// On success, the following information will be stored in the catalog entry representing the
+/// target delay rule:
+/// - The hostname or IP of the delay server
+/// - The PID of the delay server
+/// - The time the entry was claimed
+///
+/// This operation will fail if the delay rule has already been claimed.
+/// TODO Consider if it would be good to always return who claimed the rule.
+///
+/// \param[in] _comm  A pointer to a RcComm.
+/// \param[in] _input A pointer to a DelayRuleClaimInput.
+///
+/// \return An integer.
+/// \retval 0  On success.
+/// \retval <0 On failure.
+///
+/// \b Example
+/// \code{.cpp}
+/// RcComm* comm = // Our iRODS connection representing a rodsadmin user.
+///
+/// // Configure the input object for the API call.
+/// struct DelayRuleClaimInput input;
+/// memset(&input, 0, sizeof(struct DelayRuleClaimInput));
+///
+/// strncpy(input.rule_id, "12345", sizeof(DelayRuleClaimInput::rule_id) - 1);
+/// strncpy(input.execution_host, "node1.example.com", sizeof(DelayRuleClaimInput::execution_host) - 1);
+///
+/// // We're operating in the delay server, so we can use getpid().
+/// input.execution_host_pid = getpid();
+///
+/// const int ec = rc_delay_rule_claim(comm, &input);
+///
+/// if (ec < 0) {
+///     // Handle error.
+/// }
+/// \endcode
+///
+/// \since 5.0.0
+int rc_delay_rule_claim(struct RcComm* _comm, struct DelayRuleClaimInput* _input);
```

Comment:
After some discussion, we've decided to rename the APIs, etc to use `lock` instead of `claim`.

APIs
- rc_delay_rule_lock
- rc_delay_rule_unlock

GenQuery
- RULE_EXEC_LOCK_HOST
- RULE_EXEC_LOCK_HOST_PID
- RULE_EXEC_LOCK_TIME

database
- lock_host
- lock_host_pid
- lock_ts

Reply:
Done